src/java.base/share/classes/sun/security/pkcs/PKCS8Key.java

```diff
@@ -201,34 +210,41 @@ public String getFormat() {
     }
 
     /**
-     * DER-encodes this key as a byte array that can be retrieved
-     * by the {@link #getEncoded()} method.
+     * DER-encodes this key as a byte array stored inside this object
+     * and return it.
      *
-     * @exception InvalidKeyException if an encoding error occurs.
+     * @return the encoding, or null if there is an I/O error.
      */
-    private void encode() throws InvalidKeyException {
+    private synchronized byte[] getEncodedInternal() {
         if (encodedKey == null) {
             try {
                 DerOutputStream out = new DerOutputStream ();
                 DerOutputStream tmp = new DerOutputStream();
                 tmp.putInteger(V1);
                 algid.encode(tmp);
-                tmp.putOctetString(key);
+                // Write zeroes of correct length into intermediate
+                // DerOutputStream so that the real content will not
+                // appear everywhere. The actual bytes will be copied
+                // into the final encoding so they can be cleared
+                // effectively once required.
+                tmp.putOctetString(new byte[key.length]);
                 out.write(DerValue.tag_Sequence, tmp);
                 encodedKey = out.toByteArray();
+                // Copy the actual bytes
+                System.arraycopy(key, 0, encodedKey, encodedKey.length - key.length, key.length);
```

Comment:
I think this can now be updated with the newer DerValue.wrap(....) then DerValue.clear() approach?

Reply:
Good idea!